### PR TITLE
Input class

### LINKED ISSUE #20: Input reader
Want a class to read input files. The input should include the advective and diffusive constants and boundary conditions. The file should not include hard-coded model/mesh files as their location can change. But it is ok that they include specific model dimension/tag data that is paired to the model.

### PR DESCRIPTION
**Pull request by cprbucat2**

### CODE REVIEW

File: include/MAKAsolve/Input.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+struct DirichletBC {
+	int model_dim;
+	int model_tag;
+	double value;
+};
+
+class Input {
+public:
+	SolverType solver;
```
Comment (bobpaw) [inline include/MAKAsolve/Input.h]: change this variable name (it gets confusing if everything is just called solver)

File: include/MAKAsolve/Input.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+	// physics
+	double kappa;
+	double adv_dir;
+	double adv_mag;
+
+	// boundary conditions
+	std::vector<DirichletBC> dirichletBCs;
+};
+
```
Comment (bobpaw) [inline include/MAKAsolve/Input.h:28]:
```suggestion

using InputPtr = std::unique_ptr<Input>;

```

File: include/MAKAsolve/Input.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+Input readInput(const std::string& filename);
```
Comment (bobpaw) [inline include/MAKAsolve/Input.h]:
```suggestion
InputPtr readInput(const std::string& filename);
```

File: include/MAKAsolve/Input.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#include <string>
+#include <vector>
+
+namespace maka {
+
+enum class SolverType { CPU, GPU };
+
+struct DirichletBC {
+	int model_dim;
+	int model_tag;
```
Comment (bobpaw) [inline include/MAKAsolve/Input.h]:
```suggestion
	int model_dim, model_tag;
```

File: include/MAKAsolve/Input.h (hunk 3 above)
Comment (bobpaw) [reply include/MAKAsolve/Input.h]: Functions that use the `InputPtr` should take it as a const reference.

File: include/MAKAsolve/Input.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#pragma once
```
Comment (bobpaw) [inline include/MAKAsolve/Input.h]:
```suggestion
#ifndef MAKASOLVE_INPUT_H
#define MAKASOLVE_INPUT_H
```

File: include/MAKAsolve/Input.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+	double adv_dir;
+	double adv_mag;
+
+	// boundary conditions
+	std::vector<DirichletBC> dirichletBCs;
+};
+
+Input readInput(const std::string& filename);
+
+} // namespace maka
```
Comment (bobpaw) [inline include/MAKAsolve/Input.h:33]:
```suggestion
} // namespace maka

#endif // MAKASOLVE_INPUT_H
```

File: src/Input.cc [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+		else if (key == "kappa") {
+			iss >> input.kappa;
+		} else if (key == "adv_dir") {
+			iss >> input.adv_dir;
+		} else if (key == "adv_mag") {
+			iss >> input.adv_mag;
+		} else if (key == "dirichlet") {
+			DirichletBC bc;
+			iss >> bc.model_dim >> bc.model_tag >> bc.value;
+			input.dirichletBCs.push_back(bc);
```
Comment (bobpaw) [inline src/Input.cc]: I was wrong about making the map in this function -- this is perfect.

File: test/data/square-stable.rc [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# physics
+kappa 0.5
+adv_dir 0
+adv_mag 1
+
+# boundary conditions
+dirichlet 1 6 1.0
+dirichlet 1 2 0.0
```
Comment (bobpaw) [inline test/data/square-stable.rc:8]: It doesn't really make sense to specify 0 dirichlet BCs because that is already the implicit default.

File: test/data/square-stable.rc
Comment (bobpaw) [inline test/data/square-stable.rc:1]: Rename this file to like `square-stable.rc` or `square-stable.in` or something (`.rc` stands for runtime config, `.in` would be short for input).

File: src/Input.cc [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+	while (std::getline(file, line)) {
+		if (line.empty() || line[0] == '#') continue;
+
+		std::istringstream iss(line);
+		std::string key;
+		iss >> key;
+
+		if (key == "solver") {
+			std::string value;
+			std::transform(value.begin(), value.end(), value.begin(), ::tolower);
```
Comment (bobpaw) [inline src/Input.cc]: Lowkey shocked that this seems to be the idiomatic way to convert a string to lowercase. You should be able to `#include <cctype>` and then use `std::tolower` here instead of `::tolower`.

**Review by bobpaw — CHANGES_REQUESTED** [review]
Looks good overall. I requested a few changes. You should add `Input.cc` to `src/CMakeLists.txt` and add a single test case that loads the test input file you created and confirms the right values are set.

**Comment by bobpaw** [comment]
This would close #20.

**Review by bobpaw — APPROVED** [review]
